dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h

```diff
@@ -0,0 +1,359 @@
+            }, i, std::placeholders::_1));
+    }
+
+    thread_pool_.WaitForWork();
+    for (int i = 0; i < num_threads_; ++i) {
+      CUDA_CALL(cudaEventRecord(events_[i], streams_[i]));
+      CUDA_CALL(cudaStreamWaitEvent(ws->stream(), events_[i], 0));
+    }
+  }
+
+  void SampleWorker(int i, string file_name, int in_size, int tid,
+                    const uint8_t* input_data, uint8_t* output_data) {
+    EncodedImageInfo& info = output_info_[i];
+
+    if (!info.nvjpeg_support) {
+      OCVFallback(input_data, in_size, output_data, streams_[tid], file_name);
+      return;
+    }
+
+    curr_pinned_buff_[tid] = (curr_pinned_buff_[tid] + 1) % 2;
+    const int buff_idx = tid + (curr_pinned_buff_[tid] * num_threads_);
+
+    NVJPEG_CALL(nvjpegStateAttachPinnedBuffer(image_states_[i],
+                                              pinned_buffer_[buff_idx]));
+    NVJPEG_CALL(nvjpegJpegStreamParse(handle_,
+                                      static_cast<const unsigned char*>(input_data),
+                                      in_size,
+                                      false,
+                                      false,
+                                      jpeg_streams_[buff_idx]));
+
+    CUDA_CALL(cudaEventSynchronize(events_[tid]));
+
+    nvjpegStatus_t ret = nvjpegDecodeJpegHost(handle_,
+                                              image_decoders_[i],
+                                              image_states_[i],
+                                              decode_params_[i],
+                                              jpeg_streams_[buff_idx]);
+    // If image is somehow not supported try hostdecoder
+    if (ret != NVJPEG_STATUS_SUCCESS) {
+      if (ret == NVJPEG_STATUS_JPEG_NOT_SUPPORTED || ret == NVJPEG_STATUS_BAD_JPEG) {
+        info.nvjpeg_support = false;
+      } else {
+        NVJPEG_CALL_EX(ret, file_name);
+      }
+    }
+
+    if (info.nvjpeg_support) {
+      nvjpegImage_t nvjpeg_image;
+      nvjpeg_image.channel[0] = output_data;
+      nvjpeg_image.pitch[0] = NumberOfChannels(output_image_type_) * info.widths[0];
+
+      NVJPEG_CALL(nvjpegStateAttachDeviceBuffer(image_states_[i], device_buffer_[tid]));
+
+      NVJPEG_CALL(nvjpegDecodeJpegTransferToDevice(
+          handle_,
+          image_decoders_[i],
+          image_states_[i],
+          jpeg_streams_[buff_idx],
+          streams_[tid]));
+
+      // Next sample processed in this thread has to know when H2D finished
+      CUDA_CALL(cudaEventRecord(events_[tid], streams_[tid]));
+
+      NVJPEG_CALL(nvjpegDecodeJpegDevice(
+          handle_,
+          image_decoders_[i],
+          image_states_[i],
+          &nvjpeg_image,
+          streams_[tid]));
+    } else {
+      OCVFallback(input_data, in_size, output_data, streams_[tid], file_name);
+    }
+  }
+
+  bool ShouldBeHybrid(EncodedImageInfo& info) {
+    return info.widths[0] * info.heights[0] > hybrid_huffman_threshold_;
```

Comment:
I struggled for a while until came up, that this function checks, if decoding should be performed in hybrid way. Can you write a documentation, to disambiguate it?

Reply:
maybe renaming to `IsHybridHuffman` or similar will be self-documenting enough?

Reply:
Done